### DISCUSSION TITLE
[Solved but with questions!] Testcafe won’t run with yarn2

### DISCUSSION THREAD

**Discussion by sylbru** [yarnpkg/berry#2376] [Q&A] [answered]
Hi,

Testcafe doesn’t seem to run via yarn2, while it just works running the globally installed one manually. I get this output:

```
ERROR Error: testcafe-legacy-api tried to access read-file-relative, but it isn't declared in its dependencies; this makes the require call ambiguous and unsound.

Required package: read-file-relative (via "read-file-relative")
Required by: testcafe-legacy-api@npm:4.0.0 (via /data/Fichiers/Code/testcafe/.yarn/cache/testcafe-legacy-api-npm-4.0.0-ff97b69c8c-e4a891ec41.zip/node_modules/testcafe-legacy-api/lib/)
```

I can see that `read-file-relative` _is_ listed as a dependency of `testcafe`, but _not_ as a dependency of `testcafe-legacy-api` (https://yarnpkg.com/package/testcafe, https://yarnpkg.com/package/testcafe-legacy-api), which itself is a dependency of `testcafe`.

### Solution 

I added the unlisted dependency to my `package.json`:
```
yarn add -D read-file-relative
```

And on my next `yarn run testcafe` I had the similar error with another package. I ended up installing in the top-level package (my `package.json`) everything it was missing:
```
yarn add -D testcafe-hammerhead
yarn add -D @babel/core
yarn add -D @babel/runtime
yarn add -D @babel/plugin-transform-for-of
```

Now it works great. Is it just that the dependencies listing of `testcafe` and `testcafe-legacy-api` are wrong? Should I report a bug there?

**Comment by merceyz** [comment, marked as answer]
The documentation explains how to deal with undeclared dependencies here https://yarnpkg.com/getting-started/migration#fix-dependencies-with-packageextensions and here https://yarnpkg.com/getting-started/migration/#a-package-is-trying-to-access-another-package-

> Is it just that the dependencies listing of testcafe and testcafe-legacy-api are wrong? Should I report a bug there?

Yes, they lack a dependency and need to list it

**Reply by sylbru** [reply]
Thanks for pointing me to these more accurate workarounds! I’m writing a bug report to TestCafe right now.
